test/server/server_test.cc

```diff
@@ -568,17 +568,41 @@ TEST_P(ServerInstanceImplTest, BootstrapNodeWithoutAccessLog) {
                             "An admin access log path is required for a listening server.");
 }
 
+namespace {
+void bindAndListenTcpSocket(const Network::Address::InstanceConstSharedPtr& address,
+                            const Network::Socket::OptionsSharedPtr& options) {
+  auto socket = std::make_unique<Network::TcpListenSocket>(address, options, true);
+  // Some kernels erroneously allow `bind` without SO_REUSEPORT for addresses
+  // with some other socket already listening on it, see #7636.
+  if (::listen(socket->ioHandle().fd(), 1) != 0) {
+    // Mimic bind exception for the test simplicity.
+    throw Network::SocketBindException(fmt::format("cannot listen: {}", strerror(errno)), errno);
+  }
+}
+} // namespace
+
 // Test that `socket_options` field in an Admin proto is honored.
 TEST_P(ServerInstanceImplTest, BootstrapNodeWithSocketOptions) {
+  // Start Envoy instance with admin port with SO_REUSEPORT option.
   ASSERT_NO_THROW(initialize("test/server/node_bootstrap_with_admin_socket_options.yaml"));
   const auto address = server_->admin().socket().localAddress();
-  EXPECT_THAT_THROWS_MESSAGE(std::make_unique<Network::TcpListenSocket>(address, nullptr, true),
-                             EnvoyException, HasSubstr("Address already in use"));
+
+  // First attempt to bind and listen socket should fail due to the lack of SO_REUSEPORT socket
+  // options.
+  try {
```

Comment:
nit: can you use EXPECT_THROW_WITH_MESSAGE here?

Reply:
I can try to do smth like `EXPECT_THROW_WITH_MESSAGE(bindAndListenTcpSocket(address, nullptr), strerror(EADDRINUSE))`, would it be clear enough?

Reply:
Yes any variant of that is fine, there is also a regex version. Basically just trying to avoid re-implementing the construct.